fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/AddDecodeNodeForDictStringRule.java

```diff
@@ -0,0 +1,411 @@
+            return optExpression;
+        }
+
+        @Override
+        public OptExpression visitPhysicalDecode(OptExpression optExpression, DecodeContext context) {
+            context.hasChanged = false;
+            return optExpression;
+        }
+
+        @Override
+        public OptExpression visitPhysicalOlapScan(OptExpression optExpression, DecodeContext context) {
+            PhysicalOlapScanOperator scanOperator = (PhysicalOlapScanOperator) optExpression.getOp();
+            long tableId = scanOperator.getTable().getId();
+            if (context.tableIdToStringColumnIds.containsKey(scanOperator.getTable().getId())) {
+                Map<ColumnRefOperator, Column> newColRefToColumnMetaMap =
+                        Maps.newHashMap(scanOperator.getColRefToColumnMetaMap());
+
+                List<ColumnRefOperator> newOutputColumns =
+                        Lists.newArrayList(scanOperator.getOutputColumns());
+
+                for (Integer columnId : context.tableIdToStringColumnIds.get(tableId)) {
+                    ColumnRefOperator stringColumn = context.columnRefFactory.getColumnRef(columnId);
+                    if (!newOutputColumns.contains(stringColumn)) {
+                        continue;
+                    }
+
+                    // TODO(kks): support this later
+                    if (scanOperator.getPredicate() != null &&
+                            scanOperator.getPredicate().getUsedColumns().contains(columnId)) {
+                        continue;
+                    }
+
+                    ColumnRefOperator newDictColumn = context.columnRefFactory.create(
+                            stringColumn.getName(), Type.INT, stringColumn.isNullable());
+
+                    newOutputColumns.remove(stringColumn);
+                    newOutputColumns.add(newDictColumn);
+
+                    Column oldColumn = scanOperator.getColRefToColumnMetaMap().get(stringColumn);
+                    Column newColumn = new Column(oldColumn.getName(), Type.INT);
+
+                    newColRefToColumnMetaMap.remove(stringColumn);
+                    newColRefToColumnMetaMap.put(newDictColumn, newColumn);
+
+                    ColumnDict dict = IDictManager.getInstance().getGlobalDict(tableId, stringColumn.getName());
+                    // scanOperator.addGlobalDictColumns(new Pair<>(newDictColumn.getId(), dict));
+
+                    context.stringColumnIdToDictColumnIds.put(columnId, newDictColumn.getId());
+                    context.hasChanged = true;
+                }
+
+                if (context.hasChanged) {
+                    PhysicalOlapScanOperator newOlapScan = new PhysicalOlapScanOperator(
+                            scanOperator.getTable(),
+                            newOutputColumns,
+                            newColRefToColumnMetaMap,
+                            scanOperator.getDistributionSpec());
+                    newOlapScan.setSelectedIndexId(scanOperator.getSelectedIndexId());
```

Comment:
It looks like conflict with https://github.com/StarRocks/starrocks/pull/591 this pr